File: ios/Features.swift
//
//  Features.swift
//  ExpoAudioStream
//
//  Created by Arthur Breton on 23/6/2024.
//

import Foundation

public struct Features {
    var energy: Float
    var mfcc: [Float]
    var rms: Float
    var zcr: Float
    var spectralCentroid: Float
    var spectralFlatness: Float
    var spectralRollOff: Float?
    var spectralBandwidth: Float?
    var chromagram: [Float]?
    var tempo: Float?
    var hnr: Float?
}

extension Features {
    func toDictionary() -> [String: Any] {
        return [
            "energy": energy,
            "mfcc": mfcc,
            "rms": rms,
            "zcr": zcr,
            "spectralCentroid": spectralCentroid,
            "spectralFlatness": spectralFlatness,
            "spectralRollOff": spectralRollOff ?? 0,
            "spectralBandwidth": spectralBandwidth ?? 0,
            "chromagram": chromagram ?? [],
            "tempo": tempo ?? 0,
            "hnr": hnr ?? 0
        ]
    }
}

File: android/src/main/java/net/siteed/audiostream/Features.kt
package net.siteed.audiostream

import android.os.Bundle
import androidx.core.os.bundleOf

data class Features(
    val energy: Float = 0f,
    val mfcc: List<Float> = emptyList(),
    val rms: Float = 0f,
    val minAmplitude: Float = 0f,
    val maxAmplitude: Float = 0f,
    val zcr: Float = 0f,
    val spectralCentroid: Float = 0f,
    val spectralFlatness: Float = 0f,
    val spectralRollOff: Float = 0f,
    val spectralBandwidth: Float = 0f,
    val chromagram: List<Float> = emptyList(),
    val tempo: Float = 0f,
    val hnr: Float = 0f
) {
    fun toDictionary(): Map<String, Any> {
        return mapOf(
            "energy" to energy,
            "mfcc" to mfcc,
            "rms" to rms,
            "minAmplitude" to minAmplitude,
            "maxAmplitude" to maxAmplitude,
            "zcr" to zcr,
            "spectralCentroid" to spectralCentroid,
            "spectralFlatness" to spectralFlatness,
            "spectralRollOff" to spectralRollOff,
            "spectralBandwidth" to spectralBandwidth,
            "chromagram" to chromagram,
            "tempo" to tempo,
            "hnr" to hnr
        )
    }

    fun toBundle(): Bundle {
        return bundleOf(
            "energy" to energy,
            "mfcc" to mfcc,
            "rms" to rms,
            "minAmplitude" to minAmplitude,
            "maxAmplitude" to maxAmplitude,
            "zcr" to zcr,
            "spectralCentroid" to spectralCentroid,
            "spectralFlatness" to spectralFlatness,
            "spectralRollOff" to spectralRollOff,
            "spectralBandwidth" to spectralBandwidth,
            "chromagram" to chromagram,
            "tempo" to tempo,
            "hnr" to hnr
        )
    }
}
